File: services/crop-service/src/routes/health.py
"""Health check endpoints."""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import text
import redis
from ..database import get_db
from ..config import settings

router = APIRouter()


@router.get("/health")
async def health_check():
    """
    Basic health check endpoint.
    
    Returns:
        dict: Service health status
    """
    return {
        "status": "healthy",
        "service": "crop-service",
        "version": "1.0.0"
    }


@router.get("/ready")
async def readiness_check(db: Session = Depends(get_db)):
    """
    Readiness check with dependency verification.
    
    Checks:
    - Database connectivity
    - Redis connectivity
    
    Args:
        db: Database session
        
    Returns:
        dict: Service readiness status with dependency checks
    """
    checks = {
        "database": False,
        "redis": False,
    }
    
    # Check database
    try:
        db.execute(text("SELECT 1"))
        checks["database"] = True
    except Exception as e:
        checks["database"] = f"Error: {str(e)}"
    
    # Check Redis
    try:
        r = redis.from_url(settings.REDIS_URL)
        r.ping()
        checks["redis"] = True
    except Exception as e:
        checks["redis"] = f"Error: {str(e)}"
    
    all_ready = all(check is True for check in checks.values())
    
    return {
        "status": "ready" if all_ready else "not_ready",
        "service": "crop-service",
        "checks": checks
    }

File: services/govt-service/src/routes/health.ts
import { Router, Request, Response } from 'express';
import { db } from '../config/database';
import { redis } from '../config/redis';
import { logger } from '../utils/logger';

const router = Router();

/**
 * GET /health
 * Health check endpoint
 */
router.get('/', async (req: Request, res: Response) => {
  try {
    const health = {
      status: 'healthy',
      timestamp: new Date().toISOString(),
      service: 'govt-service',
      checks: {
        database: 'unknown',
        redis: 'unknown'
      }
    };

    // Check database connection
    try {
      await db.query('SELECT 1');
      health.checks.database = 'healthy';
    } catch (error) {
      logger.error('Database health check failed:', error);
      health.checks.database = 'unhealthy';
      health.status = 'degraded';
    }

    // Check Redis connection
    try {
      await redis.ping();
      health.checks.redis = 'healthy';
    } catch (error) {
      logger.error('Redis health check failed:', error);
      health.checks.redis = 'unhealthy';
      health.status = 'degraded';
    }

    const statusCode = health.status === 'healthy' ? 200 : 503;
    res.status(statusCode).json(health);
  } catch (error) {
    logger.error('Health check error:', error);
    res.status(503).json({
      status: 'unhealthy',
      timestamp: new Date().toISOString(),
      service: 'govt-service',
      error: 'Health check failed'
    });
  }
});

export default router;
